tensorflow_datasets/structured/movielens_parsing.py

```diff
@@ -0,0 +1,229 @@
+      occupation_str = occupation_index_to_label[int(occupation_index)]
+      # Combine "K-12 student" and "college/grad student" labels.
+      if occupation_str in ['K-12 student', 'college/grad student']:
+        occupation_label = 'student'
+      else:
+        occupation_label = occupation_str
+      user_info_map[user_id] = {
+          'user_gender': gender_str == 'M',
+          'bucketized_user_age': age,
+          'user_occupation_label': occupation_label,
+          'user_occupation_text': occupation_str,
+          'user_zip_code': zip_code,
+      }
+
+  ratings_file_path = os.path.join(dir_path, 'ratings.dat')
+  with tf.io.gfile.GFile(ratings_file_path, mode='rb') as ratings_file:
+    for row_num, line in enumerate(ratings_file):
+      line = codecs.decode(line, encoding='ISO-8859-1').strip()
+      user_id, movie_id, rating, timestamp = line.split('::')
+      movie_info = movie_info_map[movie_id]
+      user_info = user_info_map[user_id]
+      yield row_num, {
+          **movie_info,
+          **user_info,
+          'user_id': user_id,
+          'user_rating': rating,
+          'timestamp': timestamp,
+      }
+
+def parse_100k_movies_data(
+    dir_path: str
+) -> Iterator[Tuple[int, Dict[str, Any]]]:
+  """Parses the 100k movies data."""
+  movies_file_path = os.path.join(dir_path, 'u.item')
+  # A list for converting genre index to genre label for 100K dataset.
+  # "Children's" changed to "Children" in the list for consistency.
+  all_genre_list = [
+      'Unknown', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy',
+      'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror',
+      'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western',
+  ]
+  with tf.io.gfile.GFile(movies_file_path, mode='rb') as movies_file:
+    for row_num, line in enumerate(movies_file):
+      line = codecs.decode(line, encoding='ISO-8859-1').strip()
+      # Row format: <movie id>|<movie title>|<release date>|\
+      # <video release date>|<IMDb URL>|<19 fields for each genre>|.
+      movie_id, movie_title, _, _, _, *genre_bools = line.split('|')
+      genre_list = []
+      for index, genre_indicator in enumerate(genre_bools):
+        if genre_indicator == '1':
+          genre_list.append(all_genre_list[index])
+      yield row_num, {
+          'movie_id': movie_id,
+          'movie_title': movie_title,
+          'movie_genres': genre_list,
+      }
+
+def parse_100k_ratings_data(
+    dir_path: str
+) -> Iterator[Tuple[int, Dict[str, Any]]]:
+  """Parses the 100k ratings data."""
+  movie_info_map = {}
+  for _, movie_example in parse_100k_movies_data(dir_path):
+    movie_info_map[movie_example['movie_id']] = movie_example
+
+  users_file_path = os.path.join(dir_path, 'u.user')
+  # A list of thresholds for bucketizing user age
+  age_thresholds_list = [18, 25, 35, 45, 50, 56]
+  # A dictionary for converting 100K occupation labels to canonical labels.
```

Comment:
Add some elaboration on what is meant by "canonical" (i.e., mostly based on the 1M occupation labels, with a few exceptions).

Reply:
Added explanation in ad8a410399cac16e254f4992ec7519ff46b14a16.